packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256[] memory unlockTimeInSec_)
+    {
+        if (_subjects.length != _amounts.length) {
+            revert Staking_InvalidInputLength();
+        }
+
+        unlockTimeInSec_ = new uint256[](_subjects.length);
+
+        for (uint256 i = 0; i < _subjects.length; i++) {
+            unlockTimeInSec_[i] = _depositAndLock(
+                _subjects[i],
+                _amounts[i],
+                _lockPeriodInSec,
+                _beneficiary
+            );
+        }
+    }
+
+    /**
+     * External function to buy & lock tokens.
+     * @param _subject Subject address for which tokens are being bought & deposited.
+     * @param _depositAmount amount of moxie tokens getting deposited.
+     * @param _minReturnAmountAfterFee Slippage setting which determines minimum amount of tokens after fee.
+     * @param _lockPeriodInSec Lock period for the tokens.
+     * @return amount_ Amount of tokens bought & locked.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function buyAndLock(
+        address _subject,
+        uint256 _depositAmount,
+        uint256 _minReturnAmountAfterFee,
+        uint256 _lockPeriodInSec
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 amount_, uint256 unlockTimeInSec_)
+    {
+        moxieToken.safeTransferFrom(msg.sender, address(this), _depositAmount);
+        moxieToken.approve(address(moxieBondingCurve), _depositAmount);
+        amount_ = moxieBondingCurve.buyShares(
+            _subject,
+            _depositAmount,
+            _minReturnAmountAfterFee
+        );
+
+        (, unlockTimeInSec_) = _createLock(
+            _subject,
+            amount_,
+            _lockPeriodInSec,
+            msg.sender,
+            true
+        );
+    }
+
+    /**
+     * External function to buy & lock tokens.
+     * @param _subject Subject address for which tokens are being bought & deposited.
+     * @param _depositAmount amount of moxie tokens getting deposited.
+     * @param _minReturnAmountAfterFee Slippage setting which determines minimum amount of tokens after fee.
+     * @param _lockPeriodInSec Lock period for the tokens.
+     * @param _beneficiary Address of the beneficiary for the lock.
+     * @return amount_ Amount of tokens bought & locked.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function buyAndLockFor(
+        address _subject,
+        uint256 _depositAmount,
+        uint256 _minReturnAmountAfterFee,
+        uint256 _lockPeriodInSec,
+        address _beneficiary
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 amount_, uint256 unlockTimeInSec_)
+    {
+        moxieToken.safeTransferFrom(msg.sender, address(this), _depositAmount);
+        moxieToken.approve(address(moxieBondingCurve), _depositAmount);
+        amount_ = moxieBondingCurve.buyShares(
```

Comment:
This should be buySharesFor function

Reply:
staking contract will buy for caller & lock it. Fan tokens should be bought for staking contracts.